### DISCUSSION TITLE
Annotating `NewType` types for improved type completeness

### DISCUSSION THREAD

**Discussion by rsokl** [microsoft/pyright#4744] [Q&A] [answered]
I have [a library](https://github.com/rsokl/phantom-tensors) that exports some types created with `NewType`; e.g.,

```python
from typing import NewType as _NewType

__all__ = ["Axis", "Batch"]

Axis = _NewType("Axis", int)
Batch = _NewType("Batch", int)
```

when I run `--verifytypes` on this library, I see `error: Type is missing type annotation and could be inferred differently by type checkers` for these types. What is the proper way to annotate these? Using `TypeAlias` is not permitted due to the call syntax.

**Comment by erictraut** [comment]
Pyright special-cases calls to `NewType`, `TypeVar`, `NamedTuple` and a few others. However, it looks like you're using a call to `_NewType` (with an underscore). What is that? Is it an alias to `typing.NewType`? If it's an alias to `typing.NewType`, it should work. If it's some other method that wraps `typing.NewType`, then it won't. 

There was a bug in pyright that I fixed in 1.1.297 (published yesterday) that would affect this. Please verify that you are still having the problem in 1.1.297.

**Reply by rsokl** [reply]
Sorry, I meant to specify that `_NewType` is an alias of `typing.NewType`: `from typing import NewType as _NewType`. I updating the original post to reflect this. 

[Here](https://github.com/rsokl/phantom-tensors/blob/main/src/phantom_tensors/words.py) is the particular file in question.

> There was a bug in pyright that I fixed in 1.1.297 (published yesterday) that would affect this. Please verify that you are still having the problem in 1.1.297.

This problem arose in my attempt to upgrade from 1.1.292 to 1.1.297. [Here is the failing run](https://github.com/rsokl/phantom-tensors/actions/runs/4362317957/jobs/7627072659#step:6:70).

I am also running 1.1.297 locally and am seeing the issue on my end:

```console
$ pyright --version
pyright 1.1.297
$ pyright --ignoreexternal --verifytypes phantom_tensors | grep words
   phantom_tensors.words
phantom_tensors.words.Axis
  c:\Users\rsokl\phantom\src\phantom_tensors\words.py:35:1 - error: Type is missing type annotation and could be inferred differently by type checkers
phantom_tensors.words.Batch
  c:\Users\rsokl\phantom\src\phantom_tensors\words.py:36:1 - error: Type is missing type annotation and could be inferred differently by type checkers
(many more like these)
```

**Reply by erictraut** [reply, marked as answer]
It sounds like a bug. Could you please file a bug report with minimal repro steps and/or a minimal sample? Thanks!
